tests/cpp/test_preseg_passes.cpp

```diff
@@ -642,4 +642,491 @@ TEST_F(PresegTest, ReplaceOutput) {
+  at::Tensor at_ln_weight = at::randn({1228}, options);
+  at::Tensor at_ln_bias = at::randn({1228}, options);
+
+  auto fp32_options =
+      at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+  at::Tensor at_fp8_scale = at::randn({}, fp32_options);
+  at::Tensor at_fp8_amax = at::zeros({}, fp32_options);
+  std::vector<c10::IValue> aten_inputs = {
+      at_x,
+      at_bias,
+      at_residual,
+      at_ln_weight,
+      at_ln_bias,
+      at_fp8_scale,
+      at_fp8_amax,
+      eps};
+
+  auto args = KernelArgumentHolder::createKernelArgumentHolder(aten_inputs);
+  FusionKernelRuntime runtime(std::move(fusion_ptr), args);
+  runtime.compileFusionParallel(args);
+  auto outputs = runtime.runWithInputs(args);
+
+  // Expected result of factorAmaxReduction pass:
+  // * Two segments are created because a partial reduction and a full reduction
+  // cannot be in the same fusion.
+  // * Expect partial reduction for amax to be saved as last output of first
+  // fusion
+  // The partial amax reduction has a single reduction axis.
+  checkAmaxSegmentation(
+      runtime,
+      /*number_of_segments=*/2,
+      /*fusion_index=*/0,
+      /*number_of_outputs_in_fusion=*/3,
+      /*number_of_iterdomains=*/2,
+      /*expected_number_of_reduction_axes=*/1);
+
+  // Aten reference
+  at::Tensor at_x_cast = at_x.to(at::kFloat);
+  at::Tensor at_bias_cast = at_bias.to(at::kFloat);
+  at::Tensor at_residual_cast = at_residual.to(at::kFloat);
+  at::Tensor at_ln_weight_cast = at_ln_weight.to(at::kFloat);
+  at::Tensor at_ln_bias_cast = at_ln_bias.to(at::kFloat);
+  at::Tensor at_t1 = at_x_cast + at_bias_cast.unsqueeze(0);
+  at::Tensor at_t2 = at_t1 + at_residual_cast;
+  at::Tensor at_gamma_centered_ln_weight = at_ln_weight_cast + 1.0f;
+  at::Tensor at_t3 = at::layer_norm(
+      at_t2, {1228}, at_gamma_centered_ln_weight, at_ln_bias_cast, 1e-5);
+  at::Tensor at_t4 = at_t3 * at_fp8_scale;
+  at::Tensor at_t5 = at_t4.to(at::kBFloat16);
+  at::Tensor at_t6 = at::abs(at_t3);
+  at::Tensor at_t7 = at::max(at_t6);
+
+  testValidate(
+      runtime.fusionSegments()->completeFusion(),
+      outputs,
+      aten_inputs,
+      {at_t2, at_t5, at_t7},
+      __LINE__,
+      __FILE__);
+}
+
+TEST_F(NVFuserTest, FusionFactorAmaxHorizontalMultiplePartial_CUDA) {
+  std::unique_ptr<Fusion> fusion_ptr = std::make_unique<Fusion>();
+  Fusion& fusion = *fusion_ptr.get();
+  FusionGuard fg(fusion_ptr.get());
+
+  TensorView* tv0 = makeContigTensor(2, DataType::Float);
+  TensorView* tv1 = makeContigTensor(0, DataType::Float);
+  fusion.addInput(tv0);
+  fusion.addInput(tv1);
+
+  // Add Inner and Outer Partial Reductions together
+  TensorView* tv2 = sum(tv0, {1}, /*keepdim=*/true);
+  TensorView* tv3 = sum(tv0, {0}, /*keepdim=*/false);
+  TensorView* tv4 = add(tv2, tv3);
+  fusion.addOutput(tv4);
+
+  // Full Amax Reduction
+  TensorView* tv5 = abs(tv4);
+  TensorView* tv6 = max(tv5, {0, 1}, /*keepdim=*/false);
```

Comment:
brainstorm here.

From this patten, looks like we cannot really merge the partial reduction if there's a divergence in the reduction axes in its immediate dominating reductions.

Should we rethink this criteria:
> 3. Select upstream, partial reduction.
> Ideally, we would pick the partial reduction that lets us save the smallest partial Amax reduction. However, in this pre-segmentation pass, we do not know how the fusion could be segmented nor the size of the tensors. Given this lack of information, we select the partial reduction with shortest dependency chain with Amax reduction.

i.e. maybe we should back out from this optimization unless all of its immediate dominating reductions all share identical reduction axes?